src/gax.ts

```diff
@@ -67,27 +69,53 @@ import {toLowerCamelCase} from './util';
 
 /**
  * Per-call configurable settings for retrying upon transient failure.
+ * @implements {RetryOptionsType}
  * @typedef {Object} RetryOptions
- * @property {String[]} retryCodes
+ * @property {String[] | (function)} retryCodesOrShouldRetryFn
  * @property {BackoffSettings} backoffSettings
+ * @property {(function)} getResumptionRequestFn
  */
 export class RetryOptions {
-  retryCodes: number[];
+  retryCodesOrShouldRetryFn: number[] | ((error?: GoogleError) => boolean);
   backoffSettings: BackoffSettings;
-  constructor(retryCodes: number[], backoffSettings: BackoffSettings) {
-    this.retryCodes = retryCodes;
+  getResumptionRequestFn?: (response: any) => any;
+  constructor(
+    retryCodesOrShouldRetryFn: number[] | ((error?: GoogleError) => boolean),
+    backoffSettings: BackoffSettings,
+    getResumptionRequestFn?: (response: any) => any
+  ) {
+    this.retryCodesOrShouldRetryFn = retryCodesOrShouldRetryFn;
```

Comment:
The fact that this could be totally two different types is causing a lot of extra code to check what it is - for example, lines 250 - 252 in src/gax, lines 113 - 114 in createApiCall. Just an idea, we could maybe create two separate parameters that get filled depending on what it is - for example, `retryCodes` and `shouldRetryFunction`. We could create a helper function that sets one of these depending on what it is in the constructor, that way we don't have to check it in other places in the code. Another idea is to have a typecheck that throws an error if it's not the type you expect, and then that error creates another path (i.e., if it's a function, do something else). TL;DR: I think we should centralize the typechecking in one places rather than multiple places, and at most check if a given parameter is truthy or undefined.

Reply:
I like what you're thinking with this - I know originally I had wanted to do something similar to what you're suggesting, but iirc, I ran into issues with how to require something sometimes. Do you know if we do anything like this in any other parts of gax or in our libraries? I'd love to see an example, but if not, I'll timebox messing around with this and we can follow up about it after

Reply:
Not off the top of my head for gax, but here's an example of typechecking in repo-automation-bots: https://github.com/googleapis/repo-automation-bots/blob/5941217c7119ee0600f319979a49d78eb0b94ee6/packages/auto-approve/src/get-pr-info.ts#L23

We could also do something like:
```
let retryCodes: string[];
let shouldRetryFunction : Function;

if (Array.isArray(retryCodesOrShouldRetryFn)) {
    retryCodes = retryCodesOrShouldRetryFn;
} else if (retryCodesOrShouldRetryFn instanceof Function) {
   shouldRetryFunction = retryCodesOrShouldRetryFn;
} else {
 // error?
}
```

Reply:
Okay so I very much want this to work but I'm running into some issues. I set up some time to talk through them later today!